### PR TITLE
Implements MultiThreadResume for CpuManager

### PR DESCRIPTION
**Pull request by SuchAFuriousDeath**

### CODE REVIEW

**Comment by SuchAFuriousDeath** [comment]
There are likely some things wrong in this PR, it requires review

**Comment by ultimaweapon** [comment]
I'm not sure how `gdbstub` tell its debug target to resume. Does the target should resume when it reached `GdbStubStateMachine::Running` or it will be a job of `MultiThreadResume` implementation?

**Comment by SuchAFuriousDeath** [comment]
I think it's the job of the resume method of MultiThreadResume

**Comment by ultimaweapon** [comment]
In that case why `MultiThreadResume` is optional to implement? Is not the resuming after hitting a breakpoint should be a fundamental debug API?

**Comment by SuchAFuriousDeath** [comment]
I asked AI and it said that it's optional, because in some cases, it's not possible to resume the target (such as when inspecting a core dump)

File: gui/src/vmm/cpu/mod.rs [hunk 1]
```diff
@@ -520,19 +534,54 @@ impl<H: Hypervisor, S: Screen> ThreadExtraInfo for CpuManager<H, S> {
...
+
+        cpu.resume_action = Some(ResumeAction::Continue);
+
+        Ok(())
+    }
+
+    #[inline(always)]
+    fn support_single_step(&mut self) -> Option<MultiThreadSingleStepOps<'_, Self>> {
+        Some(self)
+    }
```
Comment (ultimaweapon) [inline gui/src/vmm/cpu/mod.rs:568]: This should requires additional works on the hypervisor (e.g. `KVM_GUESTDBG_SINGLESTEP` on KVM). I'm not sure if the other hypervisor support this. If not we cannot enable this.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: gui/src/vmm/cpu/mod.rs (hunk 1 above)
Comment (SuchAFuriousDeath) [reply gui/src/vmm/cpu/mod.rs:568]: I'll remove it for now.

**Comment by ultimaweapon** [comment]
> I asked AI and it said that it's optional, because in some cases, it's not possible to resume the target (such as when inspecting a core dump)

In that case we need to remove the resuming code when we handled `GdbStubStateMachine::Running`.

**Comment by SuchAFuriousDeath** [comment]
Not sure what exactly you mean. Is it this?

image

**Comment by ultimaweapon** [comment]
Yes, we need to remove the call to `release` here.

**Comment by SuchAFuriousDeath** [comment]
> Yes, we need to remove the call to `release` here.

Done.

File: gui/src/vmm/cpu/controller.rs [hunk 2]
```diff
@@ -30,3 +32,7 @@ impl Drop for CpuController {
         unsafe { ManuallyDrop::take(&mut self.thread).join().unwrap() };
     }
 }
+
+pub(super) enum ResumeAction {
+    Continue,
+}
```
Comment (ultimaweapon) [inline gui/src/vmm/cpu/controller.rs:38]: I don't think we need this. According to the docs of `MultiThreadResume::resume`:

```rust
    /// Any thread that wasn't explicitly resumed by a `set_resume_action_XXX`
    /// method should be resumed as though it was resumed with
    /// `set_resume_action_continue`.
```

That mean we don't even to do anything in `clear_resume_actions` and `set_resume_action_continue`.

**Review by ultimaweapon — APPROVED** [review]

**Pull request merged**